text/profiles/0237-profiles-data-model.md

```diff
@@ -0,0 +1,1015 @@
+
+Benchmarking shows that this approach is significantly more efficient in terms of CPU utilization, memory consumption and size of the resulting protobuf payload. See [Prior art and alternatives](#prior-art-and-alternatives) for more details.
+
+#### Arrays of Integers vs Arrays of Structures
+
+Another optimization technique that we use to reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload is using arrays of integers instead of arrays of structures to represent messages. This technique is used in conjunction with the referencing technique described above. Here's pseudocode that illustrates the approach. Note that this example is simplified for the sake of clarity:
+
+```json
+// normalized
+"samples": [
+  {
+    "stacktrace_index": 1,
+    "value": 100
+  }, {
+    "stacktrace_index": 2,
+    "value": 200
+  }
+],
+
+// arrays
+"stacktrace_indices": [1, 2],
+"values": [100, 200]
+```
+
+Explanation: in `normalized` representation samples are a collection of references to other messages plus a value. The standard way of representing those is to put each [Samples](#sample-structure) into a separate message, and link from [Samples](#sample-structure) to other messages. Parsing / generating such payloads creates many individual objects that runtime has to track. The second `arrays` representation puts values of the same kind into separate arrays. This reduces the size of the resulting protobuf payload and the number of objects that need to be allocated to parse / generate such payload.
+
+Benchmarking shows that this approach is significantly more efficient in terms of CPU utilization, memory consumption and size of the resulting protobuf payload. See [Prior art and alternatives](#prior-art-and-alternatives) for more details.
+
+## Trade-offs and mitigations
+
+The biggest trade-off was made between the performance characteristics of the format and it's simplicity. The emphasis was made on the performance characteristics, which resulted in a cognitively more complex format.
+
+Authors feel like the complexity is justified for the following reasons:
+
+* as presented in [Design Goals](#design-goals) section, the performance characteristics of the format are very important for the profiling signal
+* the format is not intended to be used directly by the end users, but rather by the developers of profiling systems that are used to and are expected to be able to handle the complexity. It is not more complex than other existing formats
+
+Alternative formats that are simpler to understand were considered, but they were not as efficient in terms of CPU utilization, memory consumption and size of the resulting protobuf payload. See [next chapter, Prior art and alternatives](#prior-art-and-alternatives) for more details.
+
+## Prior art and alternatives
+
+The specification presented here was heavily inspired by pprof. Multiple alternative representations were considered, including:
```

Comment:
We already chatted about this a bit in slack, but I'm bringing it up here for visibility as well: This doc implicitly assumes that pprof itself was never an option as a format, but doesn't have any discussion about it.

I like the direction of the new format, but we need to justify it. pprof has a non-trivial ecosystem associated with it, so this needs to be weight against the possible benefits of this new format (first-class trace_id/span_id, timestamps, etc.).

Reply:
I believe the main differences of the proposed format from the pprof format are:

* Sharing of the call stacks between samples.
* Sharing of label sets between samples.
* Reuse of OpenTelemetry conventions and message types.
  * `opentelemetry.proto.common.v1.KeyValue` in the core part of the format.
  * Also `opentelemetry.proto.resource.v1.Resource` and `opentelemetry.proto.resource.v1.InstrumentationScope` in the overall proto, but those are less related to what's equivalent of pprof's format.
* Per metric / sample type sample arrays. This should help make more compact profiles that have sample types that occur at unrelated locations. I don't remember WG meetings making this decision explicitly though, I think this part could be argued both ways (the downside of the current approach is that for tightly coupled sample types like "CPU time" and "sample count" we'll be duplicating the `stacktrace_indices`, `link_indices`, `attribute_set_indices`, `timestamps` arrays.
* First-class trace ID / span ID links.
* First-class timestamp support.